### PR TITLE
RGCN minibatch using homogeneous graph.

### PR DESCRIPTION
**Pull request by classicsong**
## Description
Implement RGCN minibatch using homogeneous graph interface.

#1629 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: examples/pytorch/rgcn-hetero/README.md [hunk 1]
```diff
@@ -72,7 +72,7 @@ python3 entity_classify_mb.py -d mutag --l2norm 5e-4 --n-bases 30 --testing --gp
 
 BGS: accuracy best=96.55% avg=92.41%
 ```
-python3 entity_classify_mb.py -d bgs --l2norm 5e-4 --n-bases 40 --testing --gpu 0
+python3 entity_classify_mb.py m
```
Comment (zheng-da) [inline examples/pytorch/rgcn-hetero/README.md:75]: what is this?

File: examples/pytorch/rgcn/entity_classify_mp.py [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+        self.num_bases = None if num_bases < 0 else num_bases
+        self.num_hidden_layers = num_hidden_layers
+        self.dropout = dropout
+        self.use_self_loop = use_self_loop
+        self.low_mem = low_mem
+
+        self.layers = nn.ModuleList()
+        # i2h
+        self.layers.append(RelGraphConv(
+            self.h_dim, self.h_dim, self.num_rels, "basis",
```
Comment (zheng-da) [inline examples/pytorch/rgcn/entity_classify_mp.py:81]: are the input dim and hidden dim the same?

File: examples/pytorch/rgcn/entity_classify_mp.py [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+                             batch_size=args.batch_size,
+                             collate_fn=test_sampler.sample_blocks,
+                             shuffle=False,
+                             num_workers=args.num_workers)
+
+    if n_gpus > 1:
+        dist_init_method = 'tcp://{master_ip}:{master_port}'.format(
+            master_ip='127.0.0.1', master_port='12345')
+        world_size = n_gpus
+        th.distributed.init_process_group(backend="nccl",
```
Comment (zheng-da) [inline examples/pytorch/rgcn/entity_classify_mp.py:192]: should we use sparse embeddings to train for multiple GPUs?

File: examples/pytorch/rgcn/model.py [hunk 4]
```diff
@@ -46,3 +47,93 @@ def forward(self, g, h, r, norm):
...
+                embed = nn.Parameter(th.Tensor(input_emb_size, self.embed_size))
+                nn.init.xavier_uniform_(embed, gain=nn.init.calculate_gain('relu'))
+                self.embeds[str(ntype)] = embed
+            else:
+                loc = node_tids == ntype
+                num_subnodes = node_tids[loc].shape[0]
+                self.idmap[loc] = th.arange(num_subnodes)
+                none_embed = nn.Parameter(th.Tensor(num_subnodes, self.embed_size))
+                nn.init.xavier_uniform_(none_embed, gain=nn.init.calculate_gain('relu'))
+                self.embeds[str(ntype)] = none_embed
```
Comment (zheng-da) [inline examples/pytorch/rgcn/model.py:103]: what is the difference between these two branches? and why do you want to have nodes of different types in different embedding tensors?

File: examples/pytorch/rgcn/model.py [hunk 5]
```diff
@@ -46,3 +47,93 @@ def forward(self, g, h, r, norm):
...
+        #embeds = self.embeds[str(-1)]
+        # first we get embeddings for transductive nodes
+        #tsd_idx = node_ids < embeds.shape[0]
+        #tsd_ids = node_ids[tsd_idx]
+        #embeds = embeds[tsd_ids]
+        embeds = th.empty(node_ids.shape[0], self.embed_size, device=self.dev_id)
+        for ntype in range(self.num_of_ntype):
+            if features[ntype] is not None:
+                loc = node_tids == ntype
+                embeds[loc] = features[ntype] @ self.embeds[str(ntype)]
```
Comment (zheng-da) [inline examples/pytorch/rgcn/model.py:132]: what  is the @ operation?

File: examples/pytorch/rgcn/entity_classify_mp.py [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+                                          rank=dev_id)
+
+    # node features
+    # None for one-hot feature, if not none, it should be the feature tensor.
+    node_feats = [None] * num_of_ntype
+    embed_layer = RelGraphEmbedLayer(dev_id,
+                                     g.number_of_nodes(),
+                                     node_tids,
+                                     num_of_ntype,
+                                     node_feats,
```
Comment (zheng-da) [inline examples/pytorch/rgcn/entity_classify_mp.py:204]: the code shows how to use raw input node features, but don't really use it. I'm wondering if it makes the code over complicated?

File: examples/pytorch/rgcn-hetero/README.md (hunk 1 above)
Comment (classicsong) [reply examples/pytorch/rgcn-hetero/README.md:75]: miss typing.

File: examples/pytorch/rgcn/entity_classify_mp.py (hunk 2 above)
Comment (classicsong) [reply examples/pytorch/rgcn/entity_classify_mp.py:81]: For hidden layers, yes.
The first layer is the embedlayer.

File: examples/pytorch/rgcn/entity_classify_mp.py (hunk 3 above)
Comment (classicsong) [reply examples/pytorch/rgcn/entity_classify_mp.py:192]: Yes, I think we can add this option. (through --spare_emb)

File: examples/pytorch/rgcn/model.py (hunk 4 above)
Comment (classicsong) [reply examples/pytorch/rgcn/model.py:103]: This is for nodes having initial embeddings.
I want to write another PR to include the example.

File: examples/pytorch/rgcn/model.py (hunk 5 above)
Comment (classicsong) [reply examples/pytorch/rgcn/model.py:132]: dot.

File: examples/pytorch/rgcn/entity_classify_mp.py (hunk 6 above)
Comment (classicsong) [reply examples/pytorch/rgcn/entity_classify_mp.py:204]: This should be included. But I want to split into two PR.
I am looking at ogb for an example dataset.
